azurelinuxagent/common/logcollector.py

```diff
@@ -0,0 +1,340 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Requires Python 2.6+ and Openssl 1.0+
+#
+
+import glob
+from heapq import heappush, heappop
+import logging
+import os
+import subprocess
+import zipfile
+
+from azurelinuxagent.common.conf import get_lib_dir, get_ext_log_dir
+from azurelinuxagent.common.future import ustr
+from azurelinuxagent.common.utils.fileutil import mkdir, read_file, write_file, rm_files
+
+_EXTENSION_LOG_DIR = get_ext_log_dir()
+_AGENT_LIB_DIR = get_lib_dir()
+
+_LOG_COLLECTOR_DIR = os.path.join(_AGENT_LIB_DIR, "logcollector")
+_TRUNCATED_FILES_DIR = "/var/truncated"
+
+_OUTPUT_RESULTS_FILE_PATH = os.path.join(_LOG_COLLECTOR_DIR, "results.txt")
+_COMPRESSED_ARCHIVE_PATH = os.path.join(_LOG_COLLECTOR_DIR, "logs.zip")
+
+_MUST_COLLECT_FILES = [
+    "/var/log/waagent.log",
+    os.path.join(_AGENT_LIB_DIR, "GoalState.*.xml"),
+    os.path.join(_AGENT_LIB_DIR, "ExtensionsConfig.*.xml"),
+    os.path.join(_AGENT_LIB_DIR, "HostingEnvironmentConfig.*.xml"),
+    os.path.join(_AGENT_LIB_DIR, "SharedConfig.*.xml"),
+    os.path.join(_AGENT_LIB_DIR, "*manifest.xml"),
+    os.path.join(_AGENT_LIB_DIR, "waagent_status.json"),
+    os.path.join(_AGENT_LIB_DIR, "history", "*.zip"),
+    os.path.join(_EXTENSION_LOG_DIR, "*", "*"),
+    os.path.join(_EXTENSION_LOG_DIR, "*", "*", "*"),
+    "/var/log/waagent*",
+]
+
+_FILE_SIZE_LIMIT = 30 * 1024 * 1024  # 30 MB
+_UNCOMPRESSED_ARCHIVE_SIZE_LIMIT = 150 * 1024 * 1024  # 150 MB
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class LogCollector(object):
+
+    _TRUNCATED_FILE_PREFIX = "truncated_"
+
+    def __init__(self, manifest_file_path):
+        self._manifest_file_path = manifest_file_path
+        self._must_collect_files = self._expand_must_collect_files()
+        self._create_base_dirs()
+        self._set_logger()
+
+    @staticmethod
+    def _create_base_dirs():
+        mkdir(_TRUNCATED_FILES_DIR)
+        mkdir(_LOG_COLLECTOR_DIR)
+
+    @staticmethod
+    def _set_logger():
+        _f_handler = logging.FileHandler(_OUTPUT_RESULTS_FILE_PATH, encoding="utf-8")
+        _f_format = logging.Formatter(fmt='%(asctime)s %(levelname)s %(message)s',
+                                      datefmt='%Y-%m-%d %H:%M:%S')
+        _f_handler.setFormatter(_f_format)
+        _LOGGER.addHandler(_f_handler)
+        _LOGGER.setLevel(logging.INFO)
+
+    @staticmethod
+    def _run_shell_command(command, stdout=subprocess.PIPE, output=False):
+        def format_command(cmd):
+            return " ".join(cmd) if isinstance(cmd, list) else command
+
+        def _encode_command_output(output):
```

Comment:
why is this encoding needed?

Reply:
I maintain utf-8 encoding throughout the module (when reading files and when writing to them), so I wanted to preserve this behaviour for outputs out of my control, e.g. outputs from shell commands or stack traces. Seems safe to keep it?

Reply:
But you are creating a utf8-encoded log file already. Why would this extra encoding would be needed?